<div style="border:solid green 2px; padding: 20px">

**Hello Amber,**

My name is **John Dickson** (https://hub.tripleten.com/u/13efa053) and today I'll be reviewing your project!

You’ll find specific notes inside the project file, marked green, yellow or red.


**Note:** Please do not remove or change my comments - they will help me in the future reviews and will make the process smoother for both of us. 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment</b> 
    
Such comment will mark efficient solutions and good ideas that can be used in other projects. It will also point at the document formatting, comments, and analysis. 
</div>

<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment</b> 
    
The parts marked with yellow comments indicate that there is room for optimisation. Though the correction is not necessary it is good if you implement it.
</div>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment</b> 
    
If you see such a comment, it means that there is a problem that needs to be fixed. Please note that I won't be able to accept your project until the issue is resolved.
</div>

---
    
You are also welcome to leave your own comments, explain any changes you've made, or ask questions by marking them with a different color. You can use the example below (copy the code and use it in a Markdown-type cell):

```
    
<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student’s Comment</b>

```
    
It will appear like this:
    
<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student’s Comment</b></div>
</div>

<div style="border: 5px solid red; padding: 10px; margin: 5px">
<b>Reviewer's comment V1 </b>
    
Please ensure that the project runs from start to finish before submission. 

Please also avoid using all caps, we want to try to keep things looking professional. 
</div>

<div style="border:solid green 2px; padding: 20px">
<b>Reviewer's comment v2:</b>

    
<b>Overall Feedback</b>
    
Hello Amber,

My name is **Gerardo Flores** and you can find me on the HUB as https://hub.tripleten.com/u/1949959f
    
You’ve submitted another project—great work! Your commitment to pushing through the challenges of this program is admirable.

After reviewing your submission, it is approved.
    
   
You can find my more detailed notes within your project notebook in the `Reviewer's comment v2:` section.

If you find yourself uncertain or in need of further insights, never hesitate to consult with your tutor or ask your questions here. We are here to guide and assist you.

Introduction: In this project we are using several models to assist a mobile carrier company in suggesting/predicting which plan subscribers would switch to. Aiming for a threshold of accuracy 0.75, and also checking using testing and a sanity check, we find the best model to use for this particular goal. Below, I use the models: random forest, decision tree, and logistic regression, to train and test on a percentage of the data. I later come to the conclusion that one model stands out for this goal, and explain why.

In [1]:

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

df = pd.read_csv('/datasets/users_behavior.csv')

print(df) # calls  minutes  messages   mb_used  is_ultra
print(df.head())
print(df.info())

features = df.drop('is_ultra', axis=1)
target = df['is_ultra']


      calls  minutes  messages   mb_used  is_ultra
0      40.0   311.90      83.0  19915.42         0
1      85.0   516.75      56.0  22696.96         0
2      77.0   467.66      86.0  21060.45         0
3     106.0   745.53      81.0   8437.39         1
4      66.0   418.74       1.0  14502.75         0
...     ...      ...       ...       ...       ...
3209  122.0   910.98      20.0  35124.90         1
3210   25.0   190.36       0.0   3275.61         0
3211   97.0   634.44      70.0  13974.06         0
3212   64.0   462.32      90.0  31239.78         0
3213   80.0   566.09       6.0  29480.52         1

[3214 rows x 5 columns]
   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
D

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment V1</b> 
    
Looks good.
</div>

The data has the columns calls, minutes, messages, mb_used, and is_ultra. The is_ultra column is binary, where 1 indicates an Ultra plan and 0 indicates a Smart plan. The is_ultra column will therefore be our target, because the item we are classifying is to pick the right plan. This column will be dropped from the features, and the rest of the columns will remain in the feature set to train and test the model. There are 3214 observations.

In [3]:
# train validation test 
features_temp, features_test, target_temp, target_test = train_test_split(
    features, target, test_size=0.25, random_state=12345)

features_train, features_valid, target_train, target_valid = train_test_split(
    features_temp, target_temp, test_size=0.3333, random_state=12345)

# hyperparameter-EXPLAINED BELOW
results = []

# random forest variations
for n in [10, 50, 100]:
    model = RandomForestClassifier(n_estimators=n, random_state=12345)
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    acc = accuracy_score(target_valid, predictions)
    results.append(('RandomForest', f'n_estimators={n}', acc))

# decision tree, (max_depth=None tells the model to not limit the depth of the tree)

#That means the tree can grow as deep as it wants to fit the training data.

#It’s a valid hyperparameter setting — not a mistake. 
for depth in [5, 10, None]:
    model = DecisionTreeClassifier(max_depth=depth, random_state=12345)
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    acc = accuracy_score(target_valid, predictions)
    results.append(('DecisionTree', f'max_depth={depth}', acc))

# logistic regression
model = LogisticRegression(random_state=12345, max_iter=1000)
model.fit(features_train, target_train)
predictions = model.predict(features_valid)
acc = accuracy_score(target_valid, predictions)
results.append(('LogisticRegression', 'default', acc))

#result
print("Model Comparison:")
for item in results:
    print(item[0], item[1], "Accuracy:", round(item[2], 3))


best_model = RandomForestClassifier(n_estimators=100, random_state=12345)
best_model.fit(features_train, target_train)
test_predictions = best_model.predict(features_test)
test_accuracy = accuracy_score(target_test, test_predictions)
print(f"Test set accuracy (RandomForest, n=100): {test_accuracy:.3f}")

#sanity check

most_common_class = target_train.mode()[0]  
baseline_predictions = [most_common_class] * len(target_valid)


baseline_accuracy = accuracy_score(target_valid, baseline_predictions)
print("Sanity check — baseline accuracy:", round(baseline_accuracy, 3))

Model Comparison:
RandomForest n_estimators=10 Accuracy: 0.801
RandomForest n_estimators=50 Accuracy: 0.808
RandomForest n_estimators=100 Accuracy: 0.812
DecisionTree max_depth=5 Accuracy: 0.78
DecisionTree max_depth=10 Accuracy: 0.785
DecisionTree max_depth=None Accuracy: 0.746
LogisticRegression default Accuracy: 0.738
Test set accuracy (RandomForest, n=100): 0.791
Sanity check — baseline accuracy: 0.679


<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment V1</b> 
    
Validation and test set should be around the same size. 
</div>

<div style="background-color: #d4edda; color: #155724; padding: 15px; border-radius: 4px; border: 1px solid #c3e6cb;">

<b>Reviewer's comment v2</b>
 
Well done!

Conclusion:

Split data/set size: 25% of the data was used as a test set for final evaluation. From the remaining 75%, 50% was used for training and 25% for validation. With this revised split — 25% for testing and 25% for validation — the validation and test sets are now approximately the same size, satisfying the guideline for balanced evaluation.

What models and hyperparameters were used: In the Random Forest model, the hyperparameter n_estimators was tested with values of 10, 50, and 100. The Random Forest Classifier performed best with 100 trees, reaching an accuracy of 0.812. With 50 trees, it scored 0.808, and with 10 trees, 0.801. The DecisionTreeClassifier achieved an accuracy of 0.780 with a max depth of 5, 0.785 with a depth of 10, and 0.746 with no depth limit. The LogisticRegression model, using default settings, reached an accuracy of 0.738.

Findings: The Random Forest model performed the best, achieving 81.2% accuracy on the validation set. When tested on new data, it maintained strong performance with 79.1% accuracy. The baseline model, which always predicted the most common class, achieved only 67.9% accuracy. Random Forest clearly outperformed the simpler models and the baseline, making it the best choice for this classification task.

Quality of model using test set: The best model — Random Forest Classifier with 100 trees — achieved a test set accuracy of 0.791. This means it correctly predicted whether a user is on the Ultra plan about 79.1% of the time when tested on completely new data, which is a strong and reliable result.

Sanity check: The baseline model predicted 0 for every user in the validation set and achieved an accuracy of 0.679. Any model that scores higher than this is considered better than the baseline, and Random Forest exceeded this by a significant margin.


In [ ]:
"""personal notes
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

df = pd.read_csv('/datasets/users_behavior.csv')
print(df) # calls  minutes  messages   mb_used  is_ultra
features = df.drop(['is_ultra'], axis=1)
target = df['is_ultra']

# Step 1: Split into temporary (75%) and test (25%)
features_temp, features_test, target_temp, target_test = train_test_split(
    features, target, test_size=0.25, random_state=12345)

# Step 2: Split temporary set into training (60%) and validation (15%)
features_train, features_valid, target_train, target_valid = train_test_split(
    features_temp, target_temp, test_size=0.2, random_state=12345)

# Train the model
model = RandomForestClassifier(random_state=12345)
model.fit(features_train, target_train)

# Evaluate on validation set
predictions_valid = model.predict(features_valid)
accuracy = accuracy_score(target_valid, predictions_valid)
print("Validation accuracy:", accuracy)

"""

"""new_features = pd.DataFrame(
    [
        [None, None, 2.8, 25],
        [None, None, 2.75, 25]
    ],
    columns=features.columns
model = DecisionTreeClassifier() 

#### make test so we know who's going to take it
# specify a ramdom state (number)
model = DecisionTreeClassifier(random_state=12345)

# train the model the same way as before
model.fit(features, target)
####




model.fit(features, target) #fit training, predict testing

answers = model.predict(new_features)

print(answers)"""